File: sorts/Quick.java
package sorts;

public class Quick {

	public static void sort(int[] arr) {
		sort(arr, 0, arr.length - 1);
	}

	private static void sort(int[] arr, int m, int n) {
		if (m < n) {
			int j = partition(arr, m, n);
			sort(arr, m, j - 1);
			sort(arr, j + 1, n);
		}
	}

	private static int partition(int[] arr, int m, int n) {
		int i = m + 1;
		int j = n;
		int o = arr[m];
		while (true) {
			while (i <= n && arr[i] < o) {
				i++;
			}
			while (j >= m && arr[j] > o) {
				j--;
			}
			if (i >= j) {
				break;
			}
			int t = arr[i];
			arr[i] = arr[j];
			arr[j] = t;
		}
		arr[m] = arr[j];
		arr[j] = o;
		return j;
	}

}

File: sorts/quick.go
package sorts

//QuickSort []int
func QuickSort(arr []int) {
	quickSort(arr, 0, len(arr)-1)
}

func quickSort(arr []int, m, n int) {
	if m < n {
		i := partition(arr, m, n)
		quickSort(arr, m, i-1)
		quickSort(arr, i+1, n)
	}
}

func partition(arr []int, m, n int) int {
	i, j := m+1, n
	for {
		for ; arr[i] < arr[m] && i <= n; i++ {

		}
		for ; arr[j] > arr[m] && j >= m; j-- {

		}
		if i >= j {
			break
		}
		arr[i], arr[j] = arr[j], arr[i]
	}
	arr[m], arr[j] = arr[j], arr[m]
	return j
}
